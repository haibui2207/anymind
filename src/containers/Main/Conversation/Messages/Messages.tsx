import React, { memo, useCallback, useEffect, useMemo } from "react";
import { useLazyQuery } from "@apollo/client";

import { apiUtil } from "../../../../utils";
import { IMessage, IUser } from "../../../../models";
import { useConversationContext } from "../../../../providers/ConversationProvider";

import Message from "./Message";
import useStyles from "./messages.styles";

interface IFormattedMessage extends IMessage {
  user?: IUser;
}

const Messages: React.FC<{}> = () => {
  const classes = useStyles();
  const {
    users,
    messages,
    activeUser,
    activeChannel,
    setMessages,
  } = useConversationContext();
  const [fetchMoreMessages, { called, loading, error, data }] = useLazyQuery<
    IMessage[]
  >(apiUtil.FETCH_MORE_MESSAGES);
  const formattedMessages = useMemo(
    (): IFormattedMessage[] =>
      messages.map((message: IMessage) => ({
        ...message,
        user: users.find((user: IUser) => user.id === message.userId),
      })),
    [messages]
  );

  useEffect((): void => {
    if (called && !loading) {
      if (!error && data) {
        setMessages([...data, ...messages]);
      } else {
        // TODO show error cannot get list messages
        window.alert("Cannot fetch messages");
        console.error(error);
      }
    }
  }, [messages, called, loading, error, data]);

  const handleButtonClick = useCallback((): void => {
    fetchMoreMessages({
      variables: {
        channelId: activeChannel?.id,
        messageId: messages[0].messageId,
        old: true,
      },
    });
  }, [messages, activeChannel]);

  return (
    <div className={classes.root}>
      <button
        className={classes.moreButton}
        onClick={handleButtonClick}
        disabled={loading}
      >
        {loading ? (
          <span>Loading...</span>
        ) : (
          <span>Click to load old messages</span>
        )}
      </button>
      {formattedMessages.map(
        (message: IFormattedMessage): React.ReactElement => (
          <Message
            key={message.messageId}
            {...message}
            isSender={message.userId === activeUser?.id}
          />
        )
      )}
    </div>
  );
};

export default memo(Messages);
