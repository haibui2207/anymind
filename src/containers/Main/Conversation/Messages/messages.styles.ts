import { createUseStyles } from "react-jss";

import { rem } from "../../../../utils/jss";
import COLORS from "../../../../constants/COLORS";

export default createUseStyles({
  root: {},
  moreButton: {
    border: "none",
    textAlign: "center",
    fontSize: rem(14),
    lineHeight: rem(16),
    color: COLORS.textGray,
    fontWeight: "normal",
    background: "transparent",
    display: "block",
    margin: [rem(16), "auto"],
    cursor: "pointer",
    "&:hover": { color: COLORS.textBlack },
  },
});
