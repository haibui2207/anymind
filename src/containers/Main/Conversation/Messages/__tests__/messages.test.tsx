import React from "react";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

import Messages from "../Messages";

describe("Testing Main > Conversation > Messages", (): void => {
  it("should render match snapshot", (): void => {
    const wrapper = shallow(<Messages />);
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});
