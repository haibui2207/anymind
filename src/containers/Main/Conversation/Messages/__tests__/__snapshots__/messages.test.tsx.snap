// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Main > Conversation > Messages should render match snapshot 1`] = `
<div
  className="custom-scrollbar root-0-2-1"
  id="messages-box"
/>
`;
